include/fixed_containers/struct_view.hpp

```diff
@@ -0,0 +1,869 @@
+#include "fixed_containers/fixed_set.hpp"
+#include "fixed_containers/fixed_vector.hpp"
+#include "fixed_containers/in_out.hpp"
+#include "fixed_containers/iterator_utils.hpp"
+#include "fixed_containers/memory.hpp"
+#include "fixed_containers/out.hpp"
+#include "fixed_containers/preconditions.hpp"
+#include "fixed_containers/random_access_iterator.hpp"
+#include "fixed_containers/recursive_reflection.hpp"
+#include "fixed_containers/recursive_reflection_fwd.hpp"
+#include "fixed_containers/reflection.hpp"
+#include "fixed_containers/sequence_container_checking.hpp"
+
+#include <magic_enum.hpp>
+
+#include <any>
+#include <bitset>
+#include <concepts>
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <numeric>
+#include <optional>
+#include <ranges>
+#include <string_view>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+/**
+ * Terminologies
+ *
+ * Path:
+ * A `Path` is identified with a `PathNameChain`, a sequence of struct field names that leads to a
+ * primitive field in the struct, with a caveat that when encountering an iterable, we do not
+ * include the index as part of the path, instead we use `data[:]` to represent all the elements in
+ * the iterable.
+ *
+ * `for_each_path_dfs` is a recursive function that iterates over all the paths in the
+ * struct.
+ *
+ * A `PathProperties` consists of a `StructTreeNodeMetadata` and a `FixedTensorView`, each
+ * responsible for keeping information of the data type and how to get the raw pointer to a field.
+ *
+ * `StructView`:
+ * A `StructView` is a view of a struct that is basically a mapping from a path to its
+ * `PathProperties`.
+ *
+ * `sub_struct_view_of`
+ * `sub_struct_view_of` create a view of the super struct object in the sub struct object.
+ * Currently, `sub_struct_view_of` employs a greedy strategy and updates all the indices for all
+ * paths at once. This can be improved for some usecases by using a lazy evaluation strategy, where
+ * we only update the indices when we need to. Users will specify a `ContiguousRangeSubStructView`
+ * instead of an array to denote the need of lazy evaluation.
+ *
+ * Note:
+ * `ContiguousRangeSubStructView` currently only supports flat structs.
+ *
+ *
+ * TODO: (optimization)
+ * - switch to tree data structure
+ * - support partial lazy evaluation
+ * - dynamic shape of iterable and optional
+ */
+
+/**
+ * @brief To customize bahavior of reflection and struct_view for a type
+ *
+ * Group:
+ * Metadata:
+ * - define corresponding MetadaConcept and mark it MetadataNoDefault
+ * - MetadataExtractor Specialization
+ *     - MetadataType
+ *     - CallInterface
+ * Strategy:
+ * - define corresponding StrategyConcept and mark it STRATEGY_NO_DEFAULT
+ * - ReflectionHandler Specialization
+ */
+
+namespace fixed_containers::struct_view_detail
+{
+
+using namespace fixed_containers::  // NOLINT(google-build-using-namespace)
+    recursive_reflection_detail;    // NOLINT(google-build-using-namespace)
```

Comment:
^

Reply:
call it `recursive_reflection_concept` and using namespace separately